src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+    comp->fgSnipNode(curStmt, op1);
+    comp->fgSnipNode(curStmt, op2);
+
+    // We will reuse "tree" for the loResult, which will now be of TYP_INT, and its operands
+    // will be the lo halves of op1 from above.
+    GenTree* loResult = tree;
+    loResult->SetOper(GetLoOper(loResult->OperGet()));
+    loResult->gtType = TYP_INT;
+    loResult->gtOp.gtOp1 = loOp1;
+    loResult->gtOp.gtOp2 = loOp2;
+
+    // The various halves will be correctly threaded internally. We simply need to
+    // relink them into the proper order, i.e. loOp1 is followed by loOp2, and then
+    // the loResult node.
+    // (This rethreading, and that below, are where we need to address the reverse ops case).
+    // The current order is (after snipping op1 and op2):
+    // ... loOp1-> ... hiOp1->loOp2First ... loOp2->hiOp2First ... hiOp2
+    // The order we want is:
+    // ... loOp1->loOp2First ... loOp2->loResult
+    // ... hiOp1->hiOp2First ... hiOp2->hiResult
+    // i.e. we swap hiOp1 and loOp2, and create (for now) separate loResult and hiResult trees
+    GenTree* loOp2First = hiOp1->gtNext;
+    GenTree* hiOp2First = loOp2->gtNext;
+
+    // First, we will NYI if both hiOp1 and loOp2 have side effects.
+    NYI_IF(((loOp2->gtFlags & GTF_ALL_EFFECT) != 0) && ((hiOp1->gtFlags & GTF_ALL_EFFECT) != 0),
+           "Binary long operator with non-reorderable sub expressions");
+
+    // Now, we reorder the loOps and the loResult.
+    loOp1->gtNext      = loOp2First;
+    loOp2First->gtPrev = loOp1;
+    loOp2->gtNext      = loResult;
+    loResult->gtPrev   = loOp2;
+
+    // Next, reorder the hiOps and the hiResult.
+    GenTree* hiResult = new (comp, oper) GenTreeOp(GetHiOper(oper), TYP_INT, hiOp1, hiOp2);
+    hiOp1->gtNext      = hiOp2First;
+    hiOp2First->gtPrev = hiOp1;
+    hiOp2->gtNext      = hiResult;
+    hiResult->gtPrev   = hiOp2;
+
+    if ((oper == GT_ADD) || (oper == GT_SUB))
+    {
+        if (loResult->gtOverflow())
+        {
+            hiResult->gtFlags |= GTF_OVERFLOW;
+            loResult->gtFlags &= ~GTF_OVERFLOW;
+        }
+        if (loResult->gtFlags & GTF_UNSIGNED)
+        {
+            hiResult->gtFlags |= GTF_UNSIGNED;
+        }
+    }
+
+    FinalizeDecomposition(ppTree, data, loResult, hiResult);
+}
+
+
+//------------------------------------------------------------------------
+// CreateEmbeddedFormTemp: call fgInsertEmbeddedFormTemp to replace *ppTree with
+// a new temp that is assigned to the value previously at *ppTree by inserting
+// an embedded statement. In addition, if the resulting statement actually ends
+// up being top-level, it might pull along some embedded statements that have
+// not yet been decomposed. So recursively decompose those before returning.
+//
+// Arguments:
+//    *ppTree - tree to replace with a temp.
+//
+// Return Value:
+//    The new statement that was created to create the temp.
+//
+GenTreeStmt* Decompose::CreateEmbeddedFormTemp(GenTree** ppTree)
```

Comment:
I'd just call it `CreateTemporary`, like the one in `Lowering`. The new statement isn't always embedded and being embedded is anyway a consequence of the overall design, not some special property of temporaries.